modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +327,66 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private class RemoteHostValidator {
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : line.split(" ")) {
+                if (linePart.startsWith("from=")) {
+                    for (String pattern : strip(linePart).split(",")){
+                        boolean patternIsNegated = pattern.startsWith("!");
+                        boolean patternMatchesHost = patternMatchesHost(pattern, session);
+                        if (patternIsNegated==patternMatchesHost){
+                            return false;
+                        }
+                    }
+                }
```

Comment:
Should return `false` if line contains a `from=""` stanza and no patterns match.